src/Dialect/ONNX/Transforms/Recompose.cpp

```diff
@@ -602,6 +602,57 @@ struct RecomposeQLinearMatMulFromQuantizeLinearPattern
   }
 };
 
+/// Merges nested ONNXConcatOps
+struct RecomposeConcatPattern : public OpRewritePattern<ONNXConcatOp> {
+  using OpRewritePattern<ONNXConcatOp>::OpRewritePattern;
+
+  // Helper function to check if an input is a mergeable Concat.
+  static bool isMergeableConcat(Value input, int64_t axis) {
+    auto innerConcat = input.getDefiningOp<ONNXConcatOp>();
+    if (!innerConcat)
+      return false;
+    return (innerConcat.getAxis() == axis) &&
+           (innerConcat.getResult().hasOneUse());
+  }
+
+  LogicalResult matchAndRewrite(
+      ONNXConcatOp concatOp, PatternRewriter &rewriter) const final {
+    Location loc = concatOp.getLoc();
+    auto inputs = concatOp.getOperands();
+    SmallVector<Value, 16> newInputs;
+    bool merged = false;
+
+    // Flatten nested concat nodes.
+    for (auto input : inputs) {
+      newInputs.push_back(input);
+      if (isMergeableConcat(input, concatOp.getAxis())) {
+        merged = true;
+        // Remove the nested concat and append its inputs.
+        newInputs.pop_back();
```

Comment:
push and pop would be expensive. Why don't you push `input` in the `else` part of `if(isMergeableConcat(input, concatOp.getAxis())) {`?